cpu/stm32/periph/timer.c

```diff
@@ -36,6 +36,163 @@ static inline TIM_TypeDef *dev(tim_t tim)
     return timer_config[tim].dev;
 }
 
+#ifdef MODULE_PERIPH_TIMER_PERIODIC
+
+/**
+ * @brief   Helper macro to get channel bit in timer/channel bitmap
+ */
+#define CHAN_BIT(tim, chan) (1 << chan) << (TIMER_CHANNEL_NUMOF * (tim & 1))
+
+/**
+ * @brief   Bitmap for compare channel disable after match
+ */
+static uint8_t _oneshot[(TIMER_NUMOF+1)/2];
+
+/**
+ * @brief   Bitmap to reset counter on channel match
+ */
+static uint8_t _reset_on_match[(TIMER_NUMOF+1)/2];
+
+/**
+ * @brief   Ignore one IRQ of the timer
+ */
+static uint16_t _ignore_irq;
+
+/**
+ * @brief   Clear interrupt enable after the interrupt has fired
+ */
+static inline void set_oneshot(tim_t tim, int chan)
+{
+    _oneshot[tim >> 1] |= CHAN_BIT(tim, chan);
+}
+
+/**
+ * @brief   Enable interrupt with every wrap-around of the timer
+ */
+static inline void clear_oneshot(tim_t tim, int chan)
+{
+    _oneshot[tim >> 1] &= ~CHAN_BIT(tim, chan);
+}
+
+static inline bool is_oneshot(tim_t tim, int chan)
+{
+    return _oneshot[tim >> 1] & CHAN_BIT(tim, chan);
+}
+
+/**
+ * @brief   ignore the first IRQ
+ */
+static inline void set_ignore_irq(tim_t tim)
+{
+    _ignore_irq |= 1 << tim;
+}
+
+/**
+ * @brief   don't ignore IRQ anymore
+ */
+static inline void clear_ignore_irq(tim_t tim)
+{
+    _ignore_irq &= ~(1 << tim);
+}
+
+static inline bool is_ignore_irq(tim_t tim)
+{
+    return _ignore_irq & (1 << tim);
+}
+
+static void _set_arr(tim_t tim);
+
+/**
+ * @brief   Reset the timer on channel match
+ */
+static inline void set_reset_on_match(tim_t tim, int chan)
+{
+    _reset_on_match[tim >> 1] |= CHAN_BIT(tim, chan);
+    _set_arr(tim);
+}
+
+/**
+ * @brief   Keep counting on channel match
+ */
+static inline void clear_reset_on_match(tim_t tim, int chan)
+{
+    _reset_on_match[tim >> 1] &= ~CHAN_BIT(tim, chan);
+    _set_arr(tim);
+}
+
+static inline bool is_reset_on_match(tim_t tim, int chan)
+{
+    return _reset_on_match[tim >> 1] & CHAN_BIT(tim, chan);
+}
+
+/**
+ * @brief   Set timer TOP to the lowest reset on match channel
+ */
+static void _set_arr(tim_t tim)
+{
+    uint32_t dir = dev(tim)->DIER;
+    uint32_t arr = timer_config[tim].max;
+
+    /* find the lowest reset value */
+    for (unsigned int i = 0; i < TIMER_CHANNEL_NUMOF; i++) {
+
+        /* skip disabled channels */
+        if ((dir & (TIM_DIER_CC1IE << i)) == 0) {
+            continue;
+        }
+
+        uint32_t cc = TIM_CHAN(tim, i);
+        if (is_reset_on_match(tim, i) && cc < arr) {
+            arr = cc;
+        }
+    }
+
+    dev(tim)->ARR = arr;
+}
```

Comment:
Is there a reason why `TIM_FLAG_RESET_ON_MATCH` is allowed for multiple channels? Otherwise, I wouldn't mind adding an `@pre` to the API doc that at most one channel per timer may be used with `TIM_FLAG_RESET_ON_MATCH`.

Reply:
The documentation for `TIM_FLAG_RESET_ON_MATCH` says

> When set on multiple channels, only the channel with the lowest match value will be reached.

so if you could have two channels with this flag and only the lower one would be reached.
Then if you disable the lower channel, the higher one would be used.

However, this is not used anywhere and with the Atmel timers (both on AVR & ARM) only channel 0 can be used as TOP value. 

So yea, we could extend the API documentation here.

Reply:
Maybe we could even restrict it to a certain channel, then? E.g.:

```C
#ifndef HAVE_TIMER_GET_PERIODIC_CHANNEL
static inline int timer_get_periodic_channel(tim_t tim)
{
    (void)tim;
    return 0;
}
#endif /* HAVE_TIMER_GET_PERIODIC_CHANNEL */
```

Reply:
Care to create a PR limiting the use of `TIM_FLAG_RESET_ON_MATCH` by extending the doc?

Reply:
Can I just add it here too?

Reply:
I would prefer a separate PR, as it is formally an API change.

Reply:
I moved it to a separate PR.
This doesn't change anything - `TIM_FLAG_RESET_ON_MATCH` is already documented on it's own.

Reply:
With https://github.com/RIOT-OS/RIOT/pull/14619 in, you should be able to simplify the code here.

Reply:
Alright!
Still untested though